app/controllers/preauthorize_transactions_controller.rb

```diff
@@ -1,263 +1,382 @@
-                        shipping_enabled: delivery_method == :shipping)
+    def validate_transaction_agreement(tx_params:, transaction_agreement_in_use:)
+      contract_agreed = tx_params[:contract_agreed]
 
-    if booking_data[:error].present?
-      flash[:error] = booking_data[:error]
-      return redirect_to listing_path(vprms[:listing][:id])
+      if transaction_agreement_in_use
+        if contract_agreed
+          Result::Success.new(tx_params)
+        else
+          Result::Error.new(nil, code: :agreement_missing, tx_params: tx_params)
+        end
+      else
+        Result::Success.new(tx_params)
+      end
     end
+  end
+
+  def initiate
+    validation_result = NewTransactionParams.validate(params).and_then { |params_entity|
+      tx_params = add_defaults(
+        params: params_entity,
+        shipping_enabled: listing.require_shipping_address,
+        pickup_enabled: listing.pickup_enabled)
+
+      Validator.validate_initiate_params(tx_params: tx_params,
+                                         is_booking: booking?(listing),
+                                         shipping_enabled: listing.require_shipping_address,
+                                         pickup_enabled: listing.pickup_enabled)
+    }
+
+    validation_result.on_success { |tx_params|
+      is_booking = booking?(listing)
+
+      quantity = calculate_quantity(tx_params: tx_params, is_booking: is_booking)
 
-    community_country_code = LocalizationUtils.valid_country_code(@current_community.country)
-
-    price_break_down_locals = TransactionViewUtils.price_break_down_locals({
-      booking:  true,
-      start_on: booking_data[:start_on],
-      end_on:   booking_data[:end_on],
-      duration: booking_data[:duration],
-      listing_price: vprms[:listing][:price],
-      localized_unit_type: translate_unit_from_listing(vprms[:listing]),
-      localized_selector_label: translate_selector_label_from_listing(vprms[:listing]),
-      subtotal: vprms[:subtotal],
-      shipping_price: delivery_method == :shipping ? vprms[:shipping_price] : nil,
-      total: vprms[:total_price]
-    })
-
-    render "listing_conversations/initiate", locals: {
-      preauthorize_form: PreauthorizeBookingForm.new({
-          start_on: booking_data[:start_on],
-          end_on: booking_data[:end_on]
-      }),
-      country_code: community_country_code,
-      listing: vprms[:listing],
-      delivery_method: delivery_method,
-      subtotal: vprms[:subtotal],
-      author: query_person_entity(vprms[:listing][:author_id]),
-      action_button_label: vprms[:action_button_label],
-      expiration_period: MarketplaceService::Transaction::Entity.authorization_expiration_period(vprms[:payment_type]),
-      form_action: booked_path(person_id: @current_user.id, listing_id: vprms[:listing][:id]),
-      price_break_down_locals: price_break_down_locals
+      listing_entity = ListingQuery.listing(params[:listing_id])
```

Comment:
I think ListingQuery is just extra noise at this point. The controller is already using the `listing` method everywhere so this doesn't buy anything but is an unnecessary extra thing.